### PR TITLE
RFC: Phase out log4s

### PR DESCRIPTION
**Pull request by armanbilge**
This PR replaces all uses of log4s with (hopefully) equivalent log4cats(-slf4j) APIs. The library change was proposed in https://github.com/http4s/http4s/issues/1840.

On `main` we should require an implicit `LoggerFactory` at all use-sites. I don't see how we could make that change compatibly in `series/0.23`, so we instantiate global loggers in `SyncIO` instead.

My motivation for targeting this PR at `series/0.23` is to remove all uses of log4s from http4s. This will make it easier to cross-build for Scala Native in the near future. However, it doesn't really bring any improvement and requires copious use of `unsafeRunSync()` so RFC whether or not we want to begin making this change in 0.23.x or wait until 1.x.

Thoughts?

### CODE REVIEW

**Comment by rossabaker** [comment]
Many of those are in effectful computations already and could be sequenced.

Last time I tried this, there were enough left that log4s was still useful, and could be assumed to be on the classpath anyway because almost everyone used blaze.  As that assumption changes, it's worth another try.  If the overwhelming majority of the `unsafeRunSync()` melt, great.  Where they remain, it's a bulkier log4s.

**Comment by armanbilge** [comment]
> Many of those are in effectful computations already and could be sequenced.

I'm not sure. I melted the ones I could, where `Sync` was available. Without `Sync` I don't see how we can do it?

**Review by rossabaker — COMMENTED** [review]
Oh, right.  This is why I stopped commenting before bed.

It's gross, but that just reflects that a lot of our logging is older than the concept of `LoggerFactory`.  It's also a few femtoseconds worse, but pruning dependencies is good, and I'm not volunteering to shoehorn scala-native into the non-sbt-typelevel log4s build.

I'm focused on otel instrumentation for $work, but unlike tracing and metrics, otel does not specify a novel logging API.  So surfacing `LoggerFactory` seems futureproof enough.

There's the question of whether dropping an newly unused transitive dependency is a breaking change.  I think it is, but it wouldn't need to be dragged into 1.0, or any new platforms 0.23 supports.

Seems fine to me.

**Comment by armanbilge** [comment]
> There's the question of whether dropping an newly unused transitive dependency is a breaking change.

Yes, so would I. In any case we can't drop it in 0.23, because of some "public internal" APIs that I found used in http4s-jetty or maybe one of the other satellite repos.

https://github.com/http4s/http4s/blob/3b553ccd98bb7e74b44f42aca1306976696f73e9/core/shared/src/main/scala/org/http4s/internal/InternalPlatform.scala#L22-L33

**Comment by armanbilge** [comment]
A taste of http4s Native:
https://github.com/http4s/http4s-curl

File: core/js/src/main/scala/org/http4s/Platform.scala [hunk 1]
```diff
@@ -16,7 +16,12 @@
...
 package org.http4s
 
+import cats.effect.SyncIO
+import org.typelevel.log4cats
+
 private[http4s] object Platform {
   final val isJvm = false
   final val isJs = true
+
+  lazy val loggerFactory: log4cats.LoggerFactory[SyncIO] = log4cats.noop.NoOpFactory[SyncIO]
```
[the remark this answers is not in the document]
Comment (armanbilge) [reply core/js/src/main/scala/org/http4s/Platform.scala:26]: Sorry, am I sure about what? That it should be no-op? I'm not aware of any other implementation on JS.

[the remark this answers is not in the document]
Comment (bplommer) [reply core/js/src/main/scala/org/http4s/Platform.scala:26]: Yeah, that's what I meant - oh, fair enough then

**Comment by bplommer** [comment]
Not sure how I feel about having `unsafeRunSync` everywhere - if we're going to keep doing side-effecting logging how about having a `LoggerFactory[Id]`: 
```
  lazy val loggerFactoryUnsafe: log4cats.LoggerFactory[Id] =
    loggerFactory.mapK(new FunctionK[SyncIO, Id] {
      override def apply[A](fa: SyncIO[A]): A = fa.unsafeRunSync()
    })
 ```

**Review by bplommer — APPROVED** [review]
Some quibbles but looks good to me!

**Pull request merged**